design/design-doc-uptest-improvements-and-increasing-test-coverage.md

```diff
@@ -0,0 +1,344 @@
+whether the resource is Ready or not.
+
+For this reason, a test step that checks the `Connection Details` can be added
+for some resources. These test steps can be manipulated with various hooks.
+Basically, the CLI of the relevant provider can be used here. At this point,
+it should be noted that this will be a custom step for frequently used resources
+rather than a generic step, but it can be a good point in terms of increasing
+test coverage.
+
+The `Import` and `Update` steps are managed through the annotations in the
+example manifests. It would be appropriate to consider Connection Details as a
+similar step and manage it through annotations for the resources that are
+desired to run. As a default behavior, the `Connection-Details` step will not
+run. This step can be executed if the annotation is set in the related example.
+The secret field values to be checked in the related annotation can be
+specified. For example:
+
+```yaml
+apiVersion: rds.aws.upbound.io/v1beta1
+kind: Cluster
+metadata:
+  name: example
+  annotations:
+    uptest.upbound.io/connection-details: "endpoint,master_username,port"
+    meta.upbound.io/example-id: rds/v1beta1/cluster
+spec:
+  forProvider:
+    region: us-west-1
+    engine: aurora-postgresql
+    masterUsername: cpadmin
+    autoGeneratePassword: true
+    masterPasswordSecretRef:
+      name: sample-cluster-password
+      namespace: upbound-system
+      key: password
+    skipFinalSnapshot: true
+  writeConnectionSecretToRef:
+    name: sample-rds-cluster-secret
+    namespace: upbound-system
+```
+
+Related Issue: https://github.com/upbound/official-providers-ci/issues/82
+
+### Real-Life Scenarios / Composition Testing
+
+Although Uptest has the ability to perform Claim tests, MRs are tested on a
+provider basis. However, Crossplane users often use higher-level abstractions
+such as `Compositions` and `Claims`. At the end of the day, these can be reduced
+to MRs, and it can even be said that providers are only responsible for the
+lifecycle of MRs. However, some situations are only observed when such
+higher-level abstractions are used. Therefore, it may make sense to use
+real-life scenarios in order to catch some bugs and problems in advance and
+increase test convergence.
+
+Here, imitating various user environments or creating configurations similar
+to reference platforms can be a start.
+
+### ProviderConfig Coverage
+
+As it is known, Uptest uses only `Secret` source `ProviderConfig` in its tests.
+However, crossplane providers allow many different provider configuration
+mechanisms (`IRSA`, `WebIdentitiy`, etc.). For this reason, especially the
+changes made in this context are tested manually and there is difficulty in
+preparing the environments locally. At this point, testing different
+`ProviderConfig` sources will significantly increase provider test coverage. It
+will also make things easier for this and similar cases that are difficult to
+test locally.
```

Comment:
```suggestion
Uptest only uses the `Secret` source from the `ProviderConfig` in its tests.
However, Crossplane providers allow many different provider configuration
mechanisms (`IRSA`, `WebIdentitiy`, etc.). For this reason, changes made in 
this context are tested manually and there is difficulty in
preparing the environments locally. Testing different `ProviderConfig` sources 
will significantly increase provider test coverage. It will also improve the ability
to test changes locally.
```

Reply:
Done!